# klassische Statistik

In [1]:
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
import numpy as np

plt.rc("axes.spines", top=False, right=False)
plt.rcParams["figure.figsize"] = [10, 6]
plt.rcParams["figure.dpi"] = 100

In [2]:
TEMP_PATH=os.path.join("temp")

## Import Pickle

In [3]:
train_df = pd.read_pickle(os.path.join(TEMP_PATH, "train2.pickle"))

In [4]:
train_df.head()

age         workclass  final_weight  education  education_num  \
0   39         state-gov         77516  bachelors             13   
1   50  self-emp-not-inc         83311  bachelors             13   
2   38           private        215646    hs-grad              9   
3   53           private        234721       11th              7   
4   28           private        338409  bachelors             13   

       marital_status         occupation   relationship   race     sex  \
0       never-married       adm-clerical  not-in-family  white    male   
1  married-civ-spouse    exec-managerial        husband  white    male   
2            divorced  handlers-cleaners  not-in-family  white    male   
3  married-civ-spouse  handlers-cleaners        husband  black    male   
4  married-civ-spouse     prof-specialty           wife  black  female   

   capital_gain  capital_loss  hours_per_week native_country  income  \
0          2174             0              40  united-states       0   
1             0             0              13  united-states       0   
2             0             0              40  united-states       0   
3             0             0              40  united-states       0   
4             0             0              40           cuba       0   

  hours_category age_category gain_category loss_category  native_region  
0             30           30             0             0  united-states  
1             10           40             0             0  united-states  
2             30           30             0             0  united-states  
3             30           50             0             0  united-states  
4             30           20             0             0  mittelamerika

lt Profiling Report Korrelation income mit: 
education_num 0.34 / age 0.23 / hours_per_week 0.23 / capital_gain 0.22 / capital_loss 0.15

## Statsmodels

Rückbesinnung auf Theorie:
<br>Koeffizienten sind signifikant, wenn ihr p-Wert niedrig ist zb unter 0.01; deutet darauf hin, dass es eine *statistisch* bedeutsame Beziehung zwischen dieser unabhängigen Variablen und der abhängigen Variablen gibt.

<br>In der Zusammenfassung eines Regressionsmodells zeigt der Wert unter "coef" (Koeffizient) die geschätzte Veränderung der abhängigen Variablen für eine Einheit Veränderung der unabhängigen Variablen an, vorausgesetzt, alle anderen Variablen bleiben konstant.

In [7]:
#es soll income durch Alter, Wochenstunden, Abschluss erklärt werden
model1 = smf.logit("income ~ age + hours_per_week + education_num", data=train_df)

In [8]:
results1=model1.fit()

Optimization terminated successfully.
         Current function value: 0.445274
         Iterations 7


In [9]:
print(results1.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32533
Method:                           MLE   Df Model:                            3
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.1936
Time:                        21:54:56   Log-Likelihood:                -14488.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -8.5245      0.110    -77.343      0.000      -8.740      -8.308
age                0.0469      0.001     40.451      0.000       0.045       0.049
hours_per_week     0.0428      0.001

Pseudo R nur bei 0.19; alle Koeffizienten sind hochsignifikant; education_num hat den größten Einfluss auf income mit einem Koeffizienten von 0.345

In [10]:
#es soll income durch Alter, Wochenstunden, Abschluss, Kapitalverlusten/-gewinnen erklärt werden
model2 = smf.logit("income ~ age + hours_per_week + education_num + capital_gain + capital_loss", data=train_df)

In [11]:
results2=model2.fit()

Optimization terminated successfully.
         Current function value: 0.406810
         Iterations 8


In [12]:
print(results2.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32531
Method:                           MLE   Df Model:                            5
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.2632
Time:                        21:54:56   Log-Likelihood:                -13236.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -8.3172      0.115    -72.221      0.000      -8.543      -8.091
age                0.0430      0.001     35.211      0.000       0.041       0.045
hours_per_week     0.0407      0.001

Pseudo R etwas höher mit 0.26; alle Koeffizienten sind hochsiginifikant, wobei aber der Einfluss von capital_gain und capital_loss sehr gering ist.

In [13]:
#es soll income durch Alter, Wochenstunden, Abschluss, Geschlecht und Familienstand erklärt werden
model3 = smf.logit("income ~ age + hours_per_week + education_num + sex + marital_status", data=train_df)

In [14]:
results3=model3.fit()

Optimization terminated successfully.
         Current function value: 0.371086
         Iterations 7


In [15]:
print(results3.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32526
Method:                           MLE   Df Model:                           10
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.3279
Time:                        21:54:57   Log-Likelihood:                -12074.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                              coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------
Intercept                                  -9.0177      0.139    -65.104      0.000      -9.289      -8.746
sex[T.male]                                 

In dieser Kombination wird der Koeffizient zu education_num sogar etwas größer, während die Koeffizienten age und hours_per_week an Erklärungskraft verlieren; sex male ist signifikant und hat einen eher geringen Einfluss (aber höher als die Kapital-Koeffizienten). Von den Familienstand-Kategorien sind signifikant nur die beiden married (AF und civ) und nerver-married. die married Kategorien haben einen im Vergleich hohen positiven Koeffizienten von 2,49 bzw 2,0; never-married hat einen negativen Einfluss (-0.529) auf income. pseudo R erneut gesteigert auf 0.328

In [16]:
#es soll income durch Alter, wochenstunden, Abschluss, Geschlecht, Familienstand und Beruf erklärt werden
model4 = smf.logit("income ~ age + hours_per_week + education_num + sex + marital_status + occupation", data=train_df)

In [17]:
results4=model4.fit()

Optimization terminated successfully.
         Current function value: 0.358074
         Iterations 7


In [18]:
print(results4.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32514
Method:                           MLE   Df Model:                           22
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.3515
Time:                        21:54:58   Log-Likelihood:                -11651.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                              coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------
Intercept                                  -8.2442      0.153    -53.903      0.000      -8.544      -7.944
sex[T.male]                                 

Pseudo R konnte nochmals geringfügig gesteigert werden, nun 0.352; wobei die Koeffizienten hours_per_week und education_num etwas an Erklärungskraft einbüßen, während die anderen Koeffizienten zunehmen. Von den Berufen sind craft-repair und protect-priv-serv nicht signigikant; positiven Einfluss haben exec-managerial (Führungskräfte, 0.74), prof-specialty (Fachkräfte, 0.47), sales (Verkauf, 0.2 ) und tech-support (technischer Support, 0.56), die anderen Berufe haben einen  negativen Einfluss auf income, wobei farming-fishing den höchsten negativen Wert hat (-1.29) und other-services mit -0.84 ebenfalls einen recht hohen negativen Einfluss hat.

In [19]:
#es soll income durch Alter, Wochenstunden, Abschluss, Geschlecht, Familienstand, Beruf und Herkunftsregion erklärt werden
model5 = smf.logit("income ~ age + hours_per_week + education_num + sex + marital_status + occupation + native_region", data=train_df)

In [20]:
results5=model5.fit()

Optimization terminated successfully.
         Current function value: 0.357385
         Iterations 8


In [21]:
print(results5.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32508
Method:                           MLE   Df Model:                           28
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.3527
Time:                        21:54:58   Log-Likelihood:                -11628.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                              coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------
Intercept                                  -8.4057      0.193    -43.586      0.000      -8.784      -8.028
sex[T.male]                                 

nur geringe Steigerung von pseudo R auf 0.353; die zusätzliche Variable native_region ist aber in den einzelnen Kategorien nicht signifikant, außer Mittelamerika geringfügig mit einem negativen Einfluss von -0.435 . Überlege diese Variable dann rauszulassen, da zusätzlich Verteilung sehr ungleich mit 90% "united-states"

In [22]:
#es soll income durch Alter, wochenstunden, Abschluss, Geschlecht, Familienstand, Beruf und Kapitalverluste/-gewinne erklärt werden
model6 = smf.logit("income ~ age + hours_per_week + education_num + marital_status + sex + occupation + capital_gain + capital_loss", data=train_df)

In [23]:
results6 = model6.fit()

Optimization terminated successfully.
         Current function value: 0.325235
         Iterations 8


In [24]:
print(results6.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32512
Method:                           MLE   Df Model:                           24
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.4110
Time:                        21:54:59   Log-Likelihood:                -10582.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                              coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------
Intercept                                  -8.1980      0.162    -50.584      0.000      -8.516      -7.880
marital_status[T.married-af-spouse]         

hier wurden doch noch einmal die Kapital-Variablen dazu genommen - pseudo R wird zwar nochmals etwas höher (0.411) und die Koeffizienten sind signifikant, aber sie haben quasi keinen Einfluss auf die Veränderung der abhängigen Variablen, da die Koeffizienten nur 0.0003 bzw 0.0007 sind.

Zusammenfassend lässt sich also sagen, dass die Koeffizienten age, hours_per_week,  education_num, marital_status (married-AF, married-civ und never-married), sex (male), occupation (positiv: exec-managerial, prof-specialty, tech-support // negativ vor allem farming-fishing, other-service, handlers-cleaners) sowie capital_gain und capital_loss jeweils signifikant sind und Einfluss auf die abhängige Variable income haben, wobei die Berufe und der Familienstand die größten Veränderungen beim Einkommen hervorrufen.

die Erkenntnisse in die Feature-Auswahl mit einbinden oder eher die Erkenntnisse über die Feature importance aus dem TRee???

In [26]:
model7 = smf.logit("income ~ age + hours_per_week + education_num + sex + marital_status + race", data=train_df)

In [27]:
results7 = model7.fit()

Optimization terminated successfully.
         Current function value: 0.370578
         Iterations 7


In [28]:
print(results7.summary())

                           Logit Regression Results                           
Dep. Variable:                 income   No. Observations:                32537
Model:                          Logit   Df Residuals:                    32522
Method:                           MLE   Df Model:                           14
Date:                Mon, 20 May 2024   Pseudo R-squ.:                  0.3288
Time:                        21:55:23   Log-Likelihood:                -12058.
converged:                       True   LL-Null:                       -17965.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                              coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------
Intercept                                  -9.5087      0.242    -39.283      0.000      -9.983      -9.034
sex[T.male]                                 

vom Merkmal "race" ist "white" signifikant, mit einem niedrigen positiven Koeffizienten. Ich überlege dennoch race später rauszulassen, da die Verteilung hier so ungleich ist: 85% der Fälle haben die Ausprägung "white"